### DISCUSSION TITLE
Handling pre-existing files?

### DISCUSSION THREAD

**Discussion by ceciliesdatter** [dvehrs/podget#72] [Q&A]
Just wanted to check and see if there's existing functionality of checking the library folder for existing files prior to downloading a given podcast? Or otherwise import the existing files of episodes into any corresponding system log? (Such as might be the case if switching from another podcast client.)

I'd imagine if it doesn't already exist it would be something like adding a file name check both with and without the file naming filter/mask after retrieving the next file's URL in the RSS feed?

**Comment by dvehrs** [comment]
Simple answer, yes there is a way but it is not like you describe.

TLDR HINT:  Refer to --recent option in the Podget man page.

In it's most basic form, Podget works by extracting the ENCLOSURE tags from the feed and downloading the URL that it specifies.

Over the years, we have implemented a variety of ways to have usable filenames.

First issue with names:  Depending on where the feed you are downloading comes from or the sense of humor of the person who created it, it is possible for a feed name to include characters that are not allowed on most common file systems.  When we encounter those characters, we replace them with the underscore character.  So names aren't guaranteed to be what the server thinks they are.

After that we have developed a variety of sources for filenames to make them easier for humans to understand.  We  do this because a fair number of feeds name the file in ways that make sense on the server but not to the human who downloads them.

So we have the following possible sources:
1. The actual filename at the end of the URL.  When this is formatted correctly it is often the best choice.
2. Sometimes the filename at the end is the same for every item on a feed (something like media.mp3 or podcast.mp3).  In these cases, we need to look up the URL a bit to find the unique piece that tells the server which file to download.  Often those strings are unique but completely meaningless to humans.
3.  Sometimes we need to look in the CONTENT DISPOTION tag that we can view with 'wget --server-response'.   
4. Other times, we look at the FILENAME LOCATION tag from 'wget --server-response'
5. Sometimes we accept that the filename cannot be made unique or human understandable so we prefix the name we are given with the date and time that it was downloaded.
6. Recently we added one more where we pair ENCLOSURE and TITLE tags in the RSS Feed.  Then we replace the filename with what was in the TITLE tag (with cleaning for bad characters).

Sometimes we do combine the time/date tag from step 5 with the other steps to create unique names that are easily sorted in a list.

Now if those six options weren't complicated enough, the latest version of Podget will be adding support for the MEDIA:CONTENT tag in addition to the ENCLOSURE tags for RSS feeds.  

Now, having said all of that we still need to keep a log of everything we download so we don't waste time and bandwidth needlessly downloading the same thing repeatedly.  To do that we keep the entire URL as downloaded in our logs.  It's simple and seems to work for every feed I have tested it on.  

ANSWER TO HOW WE HANDLE THIS ISSUE:

This is where the --recent option comes into play.   When we bring in a feed from another client, we run something like "podget --recent 3", and it will download the URLs from the three newest ENCLOSURE tags.  When they are successfully downloaded, the URLs are added to the 'done' log file.  The URL from every older ENCLOSURE tag is added to the 'done' log without downloading it.

Now, you only need to run Podget once with the --recent option.   Every time you run it after that it will consider the old files as already downloaded and not try to get them again.       

Was this clear or did I get lost in the weeds?

Dave

**Reply by dvehrs** [reply]
One other minor issue, when you use the --recent option it will apply to all feeds in the same serverlist file.  This may not be a problem if you have been regularly having podget download those feeds already.   However it could be and in that case it may be beneficial to put the new feed into a separate serverlist file and run podget with the --recent <count> and --serverlist <name> options when you first bring it over.  

Dave
